### PR TITLE
Add Twilio Runtime to create templates

### PR DESCRIPTION
**Pull request by stefanjudis**
## What did you implement:

Added a new template to initialize projects using [the Twilio Runtime](https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md). :) 

## How did you implement it:

Added a new `serverless.yml`, `package.json`, a function and an asset to the plugins `create` directory.

## How can we verify it:

- register at [Twilio](https://www.twilio.com/try-twilio) 
- `export TWILIO_ACCOUNT_SID=YOUR_TWILIO_ACCOUNT_SID` and `TWILIO_AUTH_TOKEN=YOUR_TWILIO_AUTH_TOKEN` – you'll find both in the [Twilio console](https://www.twilio.com/console) 
- `git clone https://github.com/pinkerton/serverless.git`
- `cd serverless`
- `npm install`
- `cd ..`
- `./serverless/bin/serverless create --template twilio-runtime --path my-project`
- `cd my-project`
- `npm install`
- `serverless deploy`
- `serverless invoke -f hello-world`
- `serverless info`
- open deployed asset from deploy logs in a browser

### Screenshots

*Create `Twilio-runtime` with local binary*

Screenshot 2019-07-29 at 10 54 40

*Run initial deploy*

Screenshot 2019-07-29 at 10 54 33

*Invoke deployed functions*

Screenshot 2019-07-29 at 10 54 27

*Retrieve invormation about the new service*

Screenshot 2019-07-29 at 10 54 18

*New asset available in Firefox*

Screenshot 2019-07-29 at 10 57 04

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [ ] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [ ] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES  
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by stefanjudis** [comment]
Hey @pmuens, thanks so much for the swift reply... :) 👌

(sorry that I missed the tests initially - I was mainly following a previous PR)

I added a unit test, but have to admit I'm a bit overwhelmed with the shell/docker magic in the integration tests. 😊

https://github.com/serverless/serverless/pull/6467/files#diff-4e5e90c6228fd48698d074241c2ba760R127

👆 Twilio functions run in `Node 8.10`... I'm not really familiar with docker and the overall setup. Should pick this image there? The tests are not really doing much in that case (if I understand that correctly) but it may be "more correct" to specify the correct node version?

**Comment by pmuens** [comment]
Great! Thanks for the update @stefanjudis :+1:

> 👆 Twilio functions run in Node 8.10... I'm not really familiar with docker and the overall setup. Should pick a matching image there? The tests are not really doing much in that case (if I understand that correctly) but it may be "more correct" to specify the correct node version?

Yes, picking the matching Node version number for the image should be sufficient. TBH we don't use such tests that often (we might even want to remove them in the future).

**Comment by stefanjudis** [comment]
@pmuens I took the closed Node image now... (I think that's good enough for now).

Is there anything else that I should have a look into?

As mentioned there's no rush on my end. :) Just want to be ready when we open the API. :)

File: lib/plugins/create/templates/twilio-runtime/handler.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const logo = require('asciiart-logo');
```
Comment (pmuens) [inline lib/plugins/create/templates/twilio-runtime/handler.js]: Where is this dependency coming from? Shouldn't this be listed in the `package.json` dependencies section? Or is this available in the Twilio Runtime? 🤔

Comment (stefanjudis) [reply lib/plugins/create/templates/twilio-runtime/handler.js]: it's defined in the `serverless.yml`. 

https://github.com/serverless/serverless/pull/6467/files#diff-c9fa620e15d789a64e35aeb136e52e5bR16-R17

I didn't want to mix this with the dependencies defined in the `package.json`. Several other parts of our tooling define the function dependencies in a separated config. For the case of serverless framework I thought this makes sense to have all the config in the `serverless.yml`. WDYT?

**Comment by stefanjudis** [comment]
> We might want to add docs, but I'm not sure if this is strictly necessary for the first step, so I'd put it on hold for the time being.

I'm assuming you're talking about [this area](https://serverless.com/framework/docs/providers/)? 

I was not sure if you folks would be up for that, but I'm happy to add docs in another PR. 🎉

File: lib/plugins/create/templates/twilio-runtime/handler.js (hunk 1 above)
Comment (pmuens) [reply lib/plugins/create/templates/twilio-runtime/handler.js]: Interesting. Haven't seen that 🤔

Thanks for clarifying :+1:

The other provider plugins are implemented in a way that one has to explicitly define them in `package.json` (or whatever runtime / package manager you're using).

I know that this is not well documented but we suggest to put everything proprietary (which is not part of the original `serverless.yml` spec) under the `custom` property. This way the Framework can be extended alter on without causing name collision.

TBH I'm fine with either way (whatever is best for Twilio), but I'd suggest to move this property to `custom`:

```yaml
custom:
  dependencies:
    asciiart-logo: '*'
```

**Comment by pmuens** [comment]
> I'm assuming you're talking about this area?
> I was not sure if you folks would be up for that, but I'm happy to add docs in another PR. 🎉

Yes, exactly. Having such docs would be really valuable!

File: lib/plugins/create/templates/twilio-runtime/handler.js (hunk 1 above)
Comment (stefanjudis) [reply lib/plugins/create/templates/twilio-runtime/handler.js]: Thanks for having another look.  @pmuens 

Your points are very interesting and bring up an additional question on my end. I had a look at the initial AWS `serverless.yml` and it turns out my proposal is different not only in the dependency part.

I defined the following different properties:

- `provider.environment` - which has a completely different meaning then in the AWS config (the reason for that is that in the twilio runtime `environment` is similar to what a `stage` in AWS is – I decided to follow naming conventions in our docs)
- `provider.config` - needed configuration token (I more or less copied that approach from the cloudflare serverless plugin implementation 🙈)
- `provider.environmentVars` - the environment variables available in the service
- `provider.dependencies` - needed npm deps which will be provided

I feel like it will be tough to match the AWS keys/properties because there are some major differences and not everything is mappable.

About the `custom` approach – I don't have strong feeling about this but I'm not sure if that's not more confusing to users to have important things in a not related `custom` property. Dependency definitions, auth credentials and environment configuration belongs on the provider level for me. 

I also looked at the other providers and it feels like there a many differences already. :)

----- 

Long story short – it feels like I won't be able to match every needed config in Twilio Runtime to AWS properties and am not sure about the `custom` property. 🙈 I'm more than happy to rename keys but for now would prefer to have everything in the `provider` (if it's really cool for you?).

**Comment by stefanjudis** [comment]
Nice! 🎉 Let's tackle docs then right after we have this here shipped. :) 🎉

**Comment by pmuens** [comment]
> Nice! 🎉 Let's tackle docs then right after we have this here shipped. :) 🎉

Sounds good! 🎉

File: lib/plugins/create/templates/twilio-runtime/handler.js (hunk 1 above)
Comment (stefanjudis) [reply lib/plugins/create/templates/twilio-runtime/handler.js]: Friendly ping @pmuens. 😊

As far as I see this is the missing part to move forward with this PR. :)

Comment (pmuens) [reply lib/plugins/create/templates/twilio-runtime/handler.js]: Hey @stefanjudis,
sorry for the slow response on my end. This somehow slipped through the cracks, so thanks for pinging me :+1:

I agree after reading your comment here. Since this is a provider plugin it's fine to have the important config parts at the `provider` level. I furthermore agree that putting things into different places will just cause confusion, so the choices you made here are absolutely right 👌

**Comment by pmuens** [comment]
Just another quick question after working with the Twilio Runtime. Does this runtime focus on Node.js? Will it be possible to use other languages as well?

I'm asking since our naming schema for Serverless templates usually is `{provider}-{runtime}` / `{provider}-{runtime}-{build-tool}`.

Given that it might be worthwhile to rename the template to `twilio-nodejs`. This way it's more future proof and clear that one has to work with Node.js.

**Comment by stefanjudis** [comment]
> Given that it might be worthwhile to rename the template to twilio-nodejs

We'll only Node and this is a very valid point. Will push an update in the next 2h and give you another ping. :) 

Thanks!!! 🎉

**Comment by stefanjudis** [comment]
@pmuens Alright. I just gave it another local try after renaming and everything still works. :) 

I think we're ready to go. 🎉🎉🎉

I'll add an example to the examples repo after we merged this PR and after the example, I'll have a look at the docs section.
